src/wp-includes/class-wp-block-metadata-registry.php

```diff
@@ -0,0 +1,228 @@
+<?php
+/**
+ * Block Metadata Registry
+ *
+ * @package WordPress
+ * @subpackage Blocks
+ * @since 6.7.0
+ */
+
+/**
+ * Class used for managing block metadata collections.
+ *
+ * The WP_Block_Metadata_Registry allows plugins to register metadata for large
+ * collections of blocks (e.g., 50-100+) using a single PHP file. This approach
+ * reduces the need to read and decode multiple `block.json` files, enhancing
+ * performance through opcode caching.
+ *
+ * @since 6.7.0
+ */
+class WP_Block_Metadata_Registry {
+
+	/**
+	 * Container for storing block metadata collections.
+	 *
+	 * Each entry maps a base path to its corresponding metadata and callback.
+	 *
+	 * @since 6.7.0
+	 * @var array<string, array<string, mixed>>
+	 */
+	private static $collections = array();
+
+	/**
+	 * Caches the last matched collection path for performance optimization.
+	 *
+	 * @since 6.7.0
+	 * @var string|null
+	 */
+	private static $last_matched_collection = null;
+
+	/**
+	 * Registers a block metadata collection.
+	 *
+	 * This method allows registering a collection of block metadata from a single
+	 * manifest file, improving performance for large sets of blocks.
+	 *
+	 * The manifest file should be a PHP file that returns an associative array, where
+	 * the keys are the block identifiers (without their namespace) and the values are
+	 * the corresponding block metadata arrays. The block identifiers must match the
+	 * parent directory name for the respective `block.json` file.
+	 *
+	 * Example manifest file structure:
+	 * ```
+	 * return array(
+	 *     'example-block' => array(
+	 *         'title' => 'Example Block',
+	 *         'category' => 'widgets',
+	 *         'icon' => 'smiley',
+	 *         // ... other block metadata
+	 *     ),
+	 *     'another-block' => array(
+	 *         'title' => 'Another Block',
+	 *         'category' => 'formatting',
+	 *         'icon' => 'star-filled',
+	 *         // ... other block metadata
+	 *     ),
+	 *     // ... more block metadata entries
+	 * );
+	 * ```
+	 *
+	 * @since 6.7.0
+	 *
+	 * @param string $path     The absolute base path for the collection ( e.g., WP_PLUGIN_DIR . '/my-plugin/blocks/' ).
+	 * @param string $manifest The absolute path to the manifest file containing the metadata collection.
+	 * @return bool True if the collection was registered successfully, false otherwise.
+	 */
+	public static function register_collection( $path, $manifest ) {
+		$path = wp_normalize_path( rtrim( $path, '/' ) );
+
+		// Check if the path is valid:
+		if ( str_starts_with( $path, wp_normalize_path( ABSPATH . WPINC ) ) ) {
+			// Core path is valid.
+		} elseif ( str_starts_with( $path, wp_normalize_path( WP_PLUGIN_DIR ) ) ) {
+			// For plugins, ensure the path is within a specific plugin directory and not the base plugin directory.
+			$plugin_dir    = wp_normalize_path( WP_PLUGIN_DIR );
```

Comment:
Repeated called to ` wp_normalize_path( WP_PLUGIN_DIR )`. Cache as variable.

Reply:
Fixed in 899bd6fe99